### PR TITLE
Alternative schema encoding

### PR DESCRIPTION
**Pull request by vil1**
This is a proposal for a new way to encode schemas.

The core encoding relies on three "essential" nodes, `One`, `:+:` and `:*:` (unit, sum and product) that should be theoretically enough to represent and types. For example, `Option[A]` is now represented as `A \/ Unit`.

The encoding also provides "extra" nodes to ease the representation of records, unions and sequences.
 
The main selling point for this encoding is that it make the whole tree homogenous: all nodes are `Schema`s as opposed with the previous encoding based on `FreeAp` in records.

This makes the use of recursion schemes (see [`cataNT`](https://github.com/scalaz/scalaz-schema/pull/26/files#diff-c6502d384f48ee6c722096ca4d701c2dR211)) rather easy.

### CODE REVIEW

File: modules/core/src/main/scala/SchemaModule.scala [hunk 1]
```diff
@@ -2,86 +2,285 @@ package scalaz
...
-    FreeAp.lift[Schema.Field[A, ?], Option[A0]](
-      Schema.Field.NonEssential(id, base, getter)
-    )
+    /**
+     * The sum of two schemas, yielding the schema for `A \/ B`
+     */
+    sealed class :+:[A, L[X] <: Schema[X], B, R[Y] <: Schema[Y]](l: => L[A], r: => R[B])
+        extends Schema[A \/ B] {
+      lazy val left: L[A]  = l
+      lazy val right: R[B] = r
```
Comment (julienrf) [inline modules/core/src/main/scala/SchemaModule.scala]: Why do we need this laziness? Is it to support recursive data types?

**Review by julienrf — COMMENTED** [review]
Interesting read.

I find it weird, though, the way you handle so-called “special cases”. We do have special cases of schemas, why do we also need `Representation` at all? Also, I find it very fragile the fact that it works by overriding incorrect default implementations.

**Comment by GrafBlutwurst** [comment]
There is an issue where `Representation` and a given `Schema` don't share the same types of term ids. in the Json case I think this leads currently to a subtle `toString` which might come and bit us later. We already have some proposals on the table to fix this see: https://gitter.im/scalaz/scalaz-schema/archives/2018/12/20

But I feel that this should probably be fixed in this PR to avoid unexpected behavior down the line

File: modules/core/src/main/scala/SchemaModule.scala (hunk 1 above)
Comment (vil1) [reply modules/core/src/main/scala/SchemaModule.scala]: Yes, that was an attempt in that direction, but I should remove it for now.

Comment (GrafBlutwurst) [reply modules/core/src/main/scala/SchemaModule.scala]: >This violates the left and right identity laws.

This is about the `Monoid[String]` right? I think that can be circumvented by folding records to `List[String]` and then generate the actual json string from that. That `Monoid` instance should be the usual lawful one.

>I find it weird, though, the way you handle so-called “special cases”

I think those are referring to `Option` and `List`. I have been mulling about this as well and in their cases they are covariant functor that have a fold which lead me down the road of thinking about generalized effect schemas but I haven't anything tangible yet. However I agree that there's more to this than we have yet.

>why do we also need Representation at all

I think it's a good way to separate what the `Schema` represents, what we want it to end up being (its `Representation`) and the required building blocks to arrive there. At least as far as I understood it.

>Also, I find it very fragile the fact that it works by overriding incorrect default implementations.

Agreed.

**Comment by vil1** [comment]
`Representation` (which is probably poorly named) is a way to separate what depends only on the target functor from what depends only on the structure of the schema.

All target functors *must* have some minimal capabilities to be able to reflect a schema's structure, but some might also *need* to perform additional operations to represent records or unions.

The former capabilities are given by the `Alt` / `Decidable` instances while the latter are bundled in the `Representation` instance. 

By default, handling a record (resp. an union) involves handling the product (resp. the sum) it contains, via the target functor's `Alt`/`Decidable` instance, and mapping the result pair through the `Iso`. Some targets, like `Gen` don't need any further processing, but others like `Encoder` needs some additional operation (ie to wrap the result in curly braces).

That's why some methods in `Representation` have a default implementation that does nothing (@julienrf is that what you refer to as incorrect default implementation?).

The goal of all that is for clients wanting to project schemas onto their own target functor to focus only on the specificity of their functor and forget about how the whole schema's structure is handled.

Regarding the conversion of `ProductTermId` and `SumTermId` (which my get renamed at some point), the current implementation of `Encoder` indeed relies on a hacky `toString`, but that's a mere shortcut (please consider that `Encoder` is a toy example that will be replaced with something more "real-world" later).

**Comment by GrafBlutwurst** [comment]
LGTM :+1:

**Review by GrafBlutwurst — APPROVED** [review]

File: modules/core/src/main/scala/Json.scala [hunk 2]
```diff
@@ -2,129 +2,34 @@ package scalaz
...
-        case JsonBool   => a => if (a) "true" else "false"
-        case JsonNull   => _ => "null"
-        case JsonNumber => a => a.shows
-      }
-    }
-  }
+  implicit final def algebra(
+    implicit primNT: Prim ~> Encoder,
+    fieldLabel: Liskov[ProductTermId, String],
+    branchLabel: Liskov[SumTermId, String]
```
Comment (emilypi) [inline modules/core/src/main/scala/Json.scala]: just use `A <~< String` for these. Much more legible.

Comment (vil1) [reply modules/core/src/main/scala/Json.scala]: Hold my beer...

Comment (vil1) [reply modules/core/src/main/scala/Json.scala]: Done

**Pull request merged**